### PR TITLE
Check enforcer event consolidation

### PR DESCRIPTION
**Pull request by mitchdenny**
This PR is an exploration of one approach to reducing the volume of requests being routed to GitHub from Check Enforcer. It works by deduping messages that are received from EventHubs as a batch. After some experimentation I am not sure that this approach will work that well (the delta reduction is not that great).

We might need to do a more stateful approach.

### CODE REVIEW

File: tools/check-enforcer/Azure.Sdk.Tools.CheckEnforcer/Handlers/CheckRunHandler.cs [hunk 1]
```diff
@@ -1,23 +1,63 @@
...
+            // only the last one since the evaluation will be the same.
+            var elligiblePayloads = from payload in payloads
+                                    where payload.CheckRun.Name != this.GlobalConfigurationProvider.GetApplicationName()
+                                    where payload.CheckRun.StartedAt < DateTimeOffset.UtcNow.Subtract(TimeSpan.FromDays(1))
+                                    where payload.CheckRun.Status == new StringEnum<CheckStatus>(CheckStatus.Completed)
+                                    select payload;
+
+            var everythingElse = from payload in elligiblePayloads
+                                 group payload by payload.CheckRun.HeadSha into payloadGroup
+                                 select payloadGroup.Last();
```
Comment (damirault) [inline tools/check-enforcer/Azure.Sdk.Tools.CheckEnforcer/Handlers/CheckRunHandler.cs:49]: Order by "completed_at" to get the most recent one? Might be useful for future caching optimizations.

File: tools/check-enforcer/Azure.Sdk.Tools.CheckEnforcer/Handlers/CheckRunHandler.cs [hunk 2]
```diff
@@ -1,23 +1,63 @@
...
+
+            var delta = elligiblePayloads.Count() - everythingElse.Count();
+
+            if (delta > 0)
+            {
+                Debugger.Break();
```
Comment (damirault) [inline tools/check-enforcer/Azure.Sdk.Tools.CheckEnforcer/Handlers/CheckRunHandler.cs:55]: Focusing on this delta I think is the most interesting thing. Are we hitting this often enough or do we really need to focus on caching the identical API calls themselves (and off box?).
[A drop in distributed cache](https://docs.microsoft.com/en-us/aspnet/core/performance/caching/distributed?view=aspnetcore-3.1#distributed-redis-cache) might be easy enough.

File: tools/check-enforcer/Azure.Sdk.Tools.CheckEnforcer/Handlers/Handler.cs [hunk 3]
```diff
@@ -161,17 +162,28 @@ private T DeserializePayload(string json)
...
+        public async Task HandleAsync(IEnumerable<string> payloads, CancellationToken cancellationToken)
         {
-            var deserializedPayload = DeserializePayload(json);
-            var installationId = deserializedPayload.Installation.Id;
-
-            var client = await this.GitHubClientProvider.GetInstallationClientAsync(installationId, cancellationToken);
-            var context = new HandlerContext<T>(deserializedPayload, client);
-
-            await HandleCoreAsync(context, cancellationToken);
+            // TODO: Handle exceptions on deserialization (too big, malformed, shouldn't
```
Comment (damirault) [inline tools/check-enforcer/Azure.Sdk.Tools.CheckEnforcer/Handlers/Handler.cs:167]: The various failure and retry cases seems like the biggest open question with this implementation. Goes directly to the retry logic in the event-hub and a per/multi message basis.

**Comment by mitchdenny** [comment]
Don't need to do this. Dropping neutral events had a massive impact. I think we've got plenty of headroom now.

**Pull request closed without merging**